### ISSUE TITLE
WatsonTcpServer.DisconnectClient : In some rare occasion, closing a connection can take up to a minute and hang there

### ISSUE THREAD

**Issue by SickLab** [jchristn/WatsonTcp#179]
I use this awesome project in an MMORPG currently running in early access on Steam.

I have many servers, and the library is running well with concurrent users well over 300 with lots of data transfer.

But sometimes, the close of a connection can hang. It forced me to thread the close connection on my server. I don't know what cause this, but sometime, a connection seems to become dead, and WatsonTcp does not deal with it the right way I think.

Just thought you should know about it to improve this issue when it happens.

**Comment by developervariety** [comment]
You should provide some examples of such issue so @jchristn can investigate.

**Comment by SickLab** [comment]
Its pretty hard to replicate, it happens in production with thousand of connection each days. All I know, is that sometimes, a connection become dead, and a Send to this connection hang. The close in WatsonTcp use a SendInternal when closing a connection. If this Send fail, the close does not complete and the dirty connection stays there indefinitely and there is no way to clean it up.

And if the server tries to send something to this connection, it hang, and it accumulate over time. A cascade of thread is building up due to this dead connection.

My guess I that during the Close connection, the SendInternal should timeout, and if it does, the Clean up of the connection should continue, not fail. This timeout should be pretty short, since we can assume the client on the other side is dead and we want to close it up anyway. Assuming the send in the close is to notify the client that we are closing the connection.

**Comment by developervariety** [comment]
Are you using the disconnect idle clients along with keep-alive features?

**Comment by SickLab** [comment]
KeepAlive is Activated.

But Settings.IdleClientTimeoutSeconds is not set, so still at 0.

I don't want to disconnect players for doing nothing, or it would be such a big value, that the problem could still occur anyway.

**Comment by jchristn** [comment]
Thanks @developervariety and hi @SickLab - have you tried disabling TCP keepalives?  There's a lot of voodoo that happens in layers beneath WatsonTcp (TCP keepalives are implemented by the OS, we merely act as a gateway to enable/disable and control them).  They've provided a mixed bag in terms of success vs failure thus far.  

Also, do you by chance monitor how many active threads there are?  There's another issue that was mentioned wherein CPU would in some circumstances drive up to 100% and remain; I have a sneaking suspicion that under some condition I did not anticipate, a disconnected client leaves a thread running rather than it being torn down as it should.  I'm curious if you notice a trend related to active threads over long periods of time.

**Comment by SickLab** [comment]
Thanks jchristn for your attention on the issue. I appreciate your work, and I like the way your library is coded. And thanks for all your work.

I will monitor the number of threads. As of right now, I am only at 40, which does not seem excessive. Will check tomorrow if it has increased and report back.

When keep alive is not activated, I have many complains from players that say they have been disconnected. It seems provider related, because many players stays connected idled without the keep alive. But keep alive, ensure that no one lost their connection.

In WatsonTcpServer.DisconnectClient, if the SendInternal takes a lot of time, which happens rarely, that’s when the problem can start to build up.

Let’s say I have a player with a "dead connection" that hang in SendInternal. Each time the server tries to send him something, a thread will pop. Many send can build up on this connection. And since even the Disconnect tries to SendAsync, it will hang up there as well. When this occurs, a Send can take up to a minute or more. It takes no time before 100 thread are on this particular connection.

Maybe we need at least a KillConnectionAsync to be able to try to Close a connection. This method could try to SendInternal, but not wait for it and perform the two other lines of code below. The client.Dispose() and the _Clients.TryRemove(...)

At least this way, the problematic connection is out of the way?

Or, a setting, that we set a SendTimeout, and when the timeout occurs, a KillConnection occur automatically.

In my case, if a Send takes more than 10 secondes, I would prefer to kill the client.

**Comment by jchristn** [comment]
Hi @SickLab thanks for your kind words.  40 doesn't seem excessive at all; I'm assuming that's inclusive of other things/tasks that your code is doing.  For reference, my ```Test.Server``` project has ~15 threads in a steady state with a couple of clients connected.

In ```DisconnectClient```, would it be helpful for you to have an option to *not* send the message to the client that it is being disconnected?  It sounds like this would bypass the send lock that is being held up by other send requests.  I believe that's what you may be proposing with ```KillConnectionAsync```

**Comment by SickLab** [comment]
Thread count has not increases today, so it’s fine on this regard, but I did not get the dead connection issue yet.

Yes, it would be helpful. Sometimes, you just want to kill a connection, not matter what.

Scenario A: When I close the server for maintenance, the close-up will fail if a connection is hanging on the send. The iteration of each disconnect will not finish, or take a very long time, because one of the 500 connection is dead.

Scenario B: If I failed to send a message to a client, I know his connection is dead, I want to be able to close him instantly. In this scenario, I am guaranteed to fall on a close that will hang.

But in each of this scenario, there are values in trying to Send the disconnect message. Its just that it should not hang there for 60 sec or more.

A Close connection with a specific SendTimeout would be great. It these scenario, I would set it to 3 secs or something like that.

Or a global SendTimeout settings? I would like a Send to no try for 60 secs or more.

Thank you very much @jchristn

**Comment by jchristn** [comment]
Got it, this all makes a lot of sense.  Please give v4.8.14.8 a try, ```DisconnectClient``` and ```DisconnectClients``` now has an optional ```bool``` in the signature that, if set to false, will bypass sending the message.  Please let me know if this fixes the issue to which you are referring.

NuGet: https://www.nuget.org/packages/WatsonTcp/4.8.14.8
Commit: https://github.com/jchristn/WatsonTcp/commit/0678d25f13caae5bca213232044037c6cbac16a5

I'm hesitant to amend one-way ```Send``` and ```SendAsync``` calls with a timeout because of the overlap with ```SendAndWait```, and, the method signatures are getting a little unruly.  I'd be open to a new set of APIs called ```SendWithTimeout``` and ```SendWithTimeoutAsync``` though, that incorporate a timeout, but do not wait on a response.  How would that sound?

**Comment by SickLab** [comment]
Thanks a lot @jchristn 

I will give it a try for sure! I am planning a server update at the start of next week. Which will include this change and any other if there is any by then.

SendWithTimeout and SendWithTimeoutAsync seems like an awesome idea.

I would use the SendWithTimeoutAsync, with a Task.ContinueWith that would call DisconnectClient(false); if the send failed. This would meet all my needs gracefully I think.

**Comment by AlonHertzog** [comment]
@jchristn
I also have an issue. I have a thread on client that checks if the client is connected, and if not, try to connect.
When I shut down the serer and rise it uo agian it works great. but when I disconnect client from server, the client don't manage to re connect.
These are the settings:
![image](https://user-images.githubusercontent.com/69188431/150209185-7e0f3e3d-20bd-497b-b5ef-5d262ee1ec9a.png)

**Comment by AlonHertzog** [comment]
@jchristn OK, more correctly this is what happening.
I hae this tcp server-client in close system of several computers. If one application stays on debug, there is a memory usage increasing on server. to avoid memory leak I check if the memry usage of the server is higher than lets say 100 mb. If so, I disconnect all the clients, and after few seconds thy suppose to reconnect by the thread.
So, what is happening that the one that is not on debug dis connect, and re connect only after I release the debug\closes this application. I am not sure why, but it is like the debug stucks something on the server.

**Comment by jchristn** [comment]
Hi @AlonHertzog are you saying that when you run your application on Debug instead of Release, you encounter this memory usage issue?  Or are you referring to another debug setting?

**Comment by AlonHertzog** [comment]
@jchristn 
No, I mean that when I run the server regular, and on client I debugging, and lets say that I am on the same palce withot forward, the memory usage of the server start to rise up. I wanted than that if I get to some memory usage, disconnect the clients to release this memory, but it is not disconnct the client that is on debug(it is disconnects another running client)

**Comment by jchristn** [comment]
Got it.  Is there any chance you could reduce this down to a code base that you could share with me to reproduce?

**Comment by AlonHertzog** [comment]
Sure. The settings of the keep alive are as I shared before. 
![image](https://user-images.githubusercontent.com/69188431/150328895-d72b3553-0f45-47e9-9b34-eddff5ddec6d.png)
That is the reconnect Task in the client side

![image](https://user-images.githubusercontent.com/69188431/150329330-9d87c462-9ad7-45ce-8330-649313a76e8c.png)
That is the Server side code of disconnecting the clients if the memory usage is high

If it is important, the client side run inside our internal nuget, that uses inside the Watson Tcp.
And BTW, thanks a lot for this great nuget, until now not a single problem, works vry good and saved us tons of time!

**Comment by jchristn** [comment]
Thank you - which version of WatsonTcp are you using?

**Comment by jchristn** [comment]
Thank you.  I put together a simple test app (see [here](https://gist.github.com/jchristn/9f2fc67a6a04c1c0e15ec00f381de23d)) in an attempt to reproduce something like what you're experiencing.  In my test, memory, CPU, and thread count remains consistent:

![image](https://user-images.githubusercontent.com/1421081/150333587-9d9553dc-bd15-45ad-a7f5-af0d3c2b1ecc.png)

However I notice:

1) The number of handles is growing over time
2) I'm getting some random exceptions when attempting to disconnect, which should not be happening:
```
1/20/2022 11:52:33 AM Failed: System.InvalidOperationException: Not connected to the server.
   at WatsonTcp.WatsonTcpClient.Disconnect() in C:\Code\Watson\WatsonTcp\WatsonTcp\WatsonTcpClient.cs:line 447
   at Test.Reconnect.Program.Main(String[] args) in C:\Code\Watson\WatsonTcp\Test.Reconnect\Program.cs:line 27
```

Could you let me know if this test was reasonably similar to what you're doing, without the obvious GUI/form app component?  I'll work on tracking these two down.

**Comment by AlonHertzog** [comment]
It is not exactly I think, since I try to disconnect all the clients from the server
. Actually during the running of the application I don't disconnect from server, only when closing, and that works well.

**Comment by AlonHertzog** [comment]
@jchristn
To be accurate, the serer have a task that sends in SensAsync every 200 ms information to all connected clients, as long as there is at least one connected client. The send to a stuck client, that is the cause to the memory usage rising

**Comment by AlonHertzog** [comment]
@jchristn
I have found a solution, I will make a PR for you next week

**Comment by jchristn** [comment]
Hi @AlonHertzog wanted to see if you had any update or if you could let me know what changes were necessary?  Cheers!

**Comment by AlonHertzhogQ** [comment]
@jchristn
Hey I actually left the old job so I didn't have the time to make PR.
What that was happening, that in the sendAsync you have sync write to headers, and than Async flush.
The problem is that while debug, this writeAsync is stuck. we made a while loop on all of the clients by iterating the list, and made SendAsync. so it made a lots of Async stuck sends while it was debugged.
It also happened on the Send I Think.
We made 2 solutions - One, because we have few clients, is to create n every new connection sepearte task for each connection to send only to him, so they won't be depend on one another, second is to use Send, but i made on the send one little change - I used AsyncWrite and than used timeout, something like WriteAsync.Wait(timeOutInMs).
On that way, is also did not stuck and the memory is flush.
I think you wanted to do something like that, you have send with timeout, but you evantually did not use that parameter of timeout.
